midend/optimizeSelectLabels.h

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2017 VMware, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http:// www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef _MIDEND_OPTIMIZESELECTLABELS_H_
+#define _MIDEND_OPTIMIZESELECTLABELS_H_
+
+#include "ir/ir.h"
+#include "frontends/common/resolveReferences/resolveReferences.h"
+#include "frontends/p4/typeMap.h"
+#include "frontends/p4/typeChecking/typeChecker.h"
+
+namespace P4 {
+
+/**
+   This class contains methods that optimize select labels.
+   If some labels overlaping or one contains the other, user will get warning
+   and code will be optimized.
+*/
+
+// This struct define node in binarry tree
+struct SelectLabel {
+    enum {ConstantLabel, RangeLabel} type;
+    union {
+      big_int constValue;
+      std::pair<big_int, big_int> rangeValue;
+    };
+
+    int base;
+    const IR::Type *typeInfo;
+    const IR::Type *rangeConstantTypeInfo;
+
+    SelectLabel(big_int value, int base, const IR::Type *typeInfo) :
```

Comment:
It looks like this is a decision tree. You should probably describe in a comment the invariants maintained by this tree - each node represents a range, the ranges are non-overlapping, etc. Calling this DecisionTree is probably a better idea, a SelectLabel does not seem an appropriate name.